stage0/src/apic.rs

```diff
@@ -0,0 +1,249 @@
+    NMI = 0b100,
+
+    /// IPI delivers an INIT request to the target local APIC(s), causing the CPU core to assume
+    /// INIT state. Trigger mode is edge-triggered, Vector must be 0x00. After INIT, target APIC
+    /// will only accept a Startup IPI, all other interrupts will be held pending.
+    Init = 0b101,
+
+    /// IPI delives a start-up request (SIPI) to the target local APIC(s) in the Destination field,
+    /// causing the core to start processing the routing whose address is specified by the Vector
+    /// field.
+    Startup = 0b110,
+}
+
+#[allow(dead_code)]
+pub enum DestinationMode {
+    // Physical destination, single local APIC ID.
+    Physical = 0,
+
+    /// Logical destination, one or more local APICs with a common destination logical ID.
+    Logical = 1,
+}
+
+pub enum Level {
+    Deassert = 0,
+    Assert = 1,
+}
+
+pub enum TriggerMode {
+    Edge = 0,
+    Level = 1,
+}
+
+#[allow(dead_code)]
+pub enum DestinationShorthand {
+    /// Destination field is required to specify the destination.
+    DestinationField = 0b00,
+
+    /// The issuing APIC is the only destination.
+    SelfOnly = 0b01,
+
+    /// The IPI is sent to all local APICs including itself (destination field = 0xFF)
+    AllInclSelf = 0b10,
+
+    /// The IPI is sent to all local APICs except itself (destination field = 0xFF)
+    AllExclSelf = 0b11,
+}
+
+impl InterruptCommandRegister {
+    pub const MSR: Msr = Msr::new(0x0000_00830);
+
+    fn write_raw(value: u64) {
+        let mut msr = Self::MSR;
+        unsafe { msr.write(value) }
+    }
+
+    pub fn write(
+        destination: u32,
+        vector: u8,
+        message_type: MessageType,
+        destination_mode: DestinationMode,
+        level: Level,
+        trigger_mode: TriggerMode,
+        destination_shorthand: DestinationShorthand,
+    ) {
+        let mut value: u64 = (destination as u64) << 32;
+        value |= (destination_shorthand as u64) << 18;
+        value |= (trigger_mode as u64) << 15;
+        value |= (level as u64) << 14;
+        value |= (destination_mode as u64) << 11;
+        value |= (message_type as u64) << 8;
+        value |= vector as u64;
+        Self::write_raw(value)
+    }
+}
+
+pub struct ErrorStatusRegister;
+
+impl ErrorStatusRegister {
+    pub const MSR: Msr = Msr::new(0x0000_00828);
+
+    fn write_raw(value: u64) {
+        let mut msr = Self::MSR;
+        unsafe { msr.write(value) }
+    }
+
+    pub fn clear() {
+        Self::write_raw(0);
+    }
+}
+
+pub struct Lapic {}
+
+impl Lapic {
+    pub fn enable() -> Result<Self, &'static str> {
+        // Safety: the CPUs we support are new enough to support CPUID.
+        let result = unsafe { __cpuid(0x0000_0001) };
+        if result.ecx & (1 << 21) == 0 {
+            return Err("x2APIC not supported");
+        }
+        // See Section 16.9 in the AMD64 Architecture Programmer's Manual, Volume 2 for explanation
+        // of the initialization procedure.
```

Comment:
i really wish they'd publish these as website, not just a pdf. would make references and looking them up a lot easier

Reply:
Feh, you should be happy that you don't have to go to the departmental library to requisition said books in print :)